### ISSUE TITLE
(e)DP doesn't work on some systems

### ISSUE THREAD

**Issue by STEFnr1** [ChefKissInc/NootedRed#52]
**Describe the bug**
The system boots with the internal display dysfunctional. Upon inspection of the kernel dmesg, it is clear that the issue is that the link training failed.

**To Reproduce**
Exact reproduction environment unknown

**Expected behavior**
Display turns off then back on, with the Apple logo visible on screen.

**Screenshots**
-Picture of black screen here-

**System Information:**
 - macOS Version: 11.7.6
 - CPU: AMD Ryzen 7 3750H with Radeon Vega Mobile Gfx
 - iGPU: AMD Radeon(TM) RX Vega 10 Graphics

**Additional context**
Related dmesg: http://0x0.st/H-i5.txt
```
ASSERT FAILED : result == DDC_RESULT_SUCESSFULL.
ASSERT LOCATION : void dp_set_panel_mode(struct dc_link *, enum dp_panel_mode) :: /Volumes/Code/p4/GoldenG/OpenGL/GLBuild/AmdRadeonLibraries/src/Latest/core/dal/dc/core/dc_link_dp.c:5062 .
[5:0:0] [DAL][EVENT][link:0] >>> type:"Link Training End", status: FAIL, lane_status: 0.
[5:0:0] [DAL][EVENT][link:0] <<< type:"Link Training End", err: kIOReturnSuccess.
```

**Comment by ChefKissInc** [comment]
@STEFnr1 This issue is known internally, but we don’t know the cause yet. 
> This may be cause by the fact that in Linux the main display driver is the NVIDIA GTX 1650
Proof? Have you tried disabling the dGPU to see if that’s actually the case?

**Comment by STEFnr1** [comment]
Disabled dGPU, but even with the dGPU enabled it still turns off the display. I know the issue is known, but it's better to have track of issues in one place.

**Comment by ChefKissInc** [comment]
@STEFnr1 I meant on linux

**Comment by STEFnr1** [comment]
I'm not a pro in Linux. How do I do that? I am booting using the live environment (or however you write it xD)

**Comment by ChefKissInc** [comment]
@STEFnr1 I’m not a linux user either…sooooo

**Comment by ChefKissInc** [comment]
Also, I think there is an instance of this issue on someone with no dGPU, it might be some kind of edge case not implemented in AMDRadeonX6000Framebuffer (it is essentially a macOS port of Display Core, so comparing the logic is mostly trivial)

**Comment by ExtremeXT** [comment]
@STEFnr1 Boolean.

**Comment by ChefKissInc** [comment]
You can just put -wegnoegpu in your boot args. But that's not gonna help with the link training issue.

**Comment by STEFnr1** [comment]
![20230525_001112](https://github.com/NootInc/NootedRed/assets/83458848/e0b96774-2dc2-4c95-98a5-03890956029b)
Without NRed, the integrated display is detected, so the other display is maybe one emulated by NVIDIA..?

**Comment by STEFnr1** [comment]
![Screenshot_20230526_134216_AnyDesk](https://github.com/NootInc/NootedRed/assets/83458848/2fd5bce1-a59a-403b-bbfc-367a01ef913b)

**Comment by STEFnr1** [comment]
Still no display

**Comment by ChefKissInc** [comment]
@jalavoui Like I said in https://github.com/NootInc/NootedRed/issues/52#issuecomment-1556293515, there's someone without a dGPU, and experiencing the same issue. The black screen is because of an error caused while doing the link training, as per the title

**Comment by ChefKissInc** [comment]
Related dmesg: http://0x0.st/H-i5.txt
```
ASSERT FAILED : result == DDC_RESULT_SUCESSFULL.
ASSERT LOCATION : void dp_set_panel_mode(struct dc_link *, enum dp_panel_mode) :: /Volumes/Code/p4/GoldenG/OpenGL/GLBuild/AmdRadeonLibraries/src/Latest/core/dal/dc/core/dc_link_dp.c:5062 .
[5:0:0] [DAL][EVENT][link:0] >>> type:"Link Training End", status: FAIL, lane_status: 0.
[5:0:0] [DAL][EVENT][link:0] <<< type:"Link Training End", err: kIOReturnSuccess.
```

**Comment by Alex-Zhangzhipeng** [comment]
Now, something strange really happen. I plugged in both video connectors, and the DP interface showed that the secondary screen lit up successfully. After sleeping, I unplugged the HDMI, and after waking up, DP can still be used.
<img width="356" alt="image" src="https://github.com/NootInc/NootedRed/assets/64397573/400d628c-2d76-4f7b-b222-db418b352430">
@ChefKissInc

**Comment by ChefKissInc** [comment]
@Alex-Zhangzhipeng Another user on our Telegram group reported this too. It’s good to know it’s consistent at the very least

**Comment by Alex-Zhangzhipeng** [comment]
Now I reboot , and the DP is still fine , what a good news!
<img width="1015" alt="截屏2023-06-20 15 56 22" src="https://github.com/NootInc/NootedRed/assets/64397573/14221b68-5b9f-4f6a-99da-220842f91941">
@ChefKissInc

**Comment by ChefKissInc** [comment]
@Alex-Zhangzhipeng Do a cold boot

**Comment by Alex-Zhangzhipeng** [comment]
ok, I will try. @ChefKissInc

**Comment by ChefKissInc** [comment]
What if you boot into another OS first?

**Comment by Alex-Zhangzhipeng** [comment]
Whether it is re-entering from the BIOS or selecting in the OC interface, the screen will not be displayed. @ChefKissInc

**Comment by ChefKissInc** [comment]
@Alex-Zhangzhipeng Sorry, I didn’t get that. Could you rephrase?

**Comment by Alex-Zhangzhipeng** [comment]
emmmm , sorry English is not my first language. At first , I change the boot device in the bios to start the windows. Then I use OpenCore to start windows. Both of these, make my computer display a black screen @ChefKissInc

**Comment by Alex-Zhangzhipeng** [comment]
Or it's better to use picture to explain it, haha..
![5a7c6412862721d02ff6415eb4076386](https://github.com/NootInc/NootedRed/assets/64397573/611b0ccc-ecc6-4c95-b69e-37660b03afc1)
![961d20aea25ffc5a72245b40e2fe3cf6](https://github.com/NootInc/NootedRed/assets/64397573/9749be57-87e6-4ccb-a085-e2559feb9923)

![d9fdd6c20e9657aa5ae51adbf75e73b3](https://github.com/NootInc/NootedRed/assets/64397573/5a0527b7-a260-46de-92a8-ecf7b5d941b4)

**Comment by ChefKissInc** [comment]
So Windows has a black screen now?

**Comment by Alex-Zhangzhipeng** [comment]
no, window is good , but when I return to Mac , Mac is black.

**Comment by ChefKissInc** [comment]
@Alex-Zhangzhipeng Alright. Install the DebugEnhancer kext along with adding `-nreddbg` to your boot args. Then, do the trick that gets macOS to display an image on your internal display. Finally, open the terminal and run the following 2 commands:
`sudo /System/Library/Extensions/AppleGraphicsControl.kext/Contents/MacOS/AGDCDiagnose | curl --progress-bar -Ffile=@- https://0x0.st`
`sudo dmesg | curl --progress-bar -Ffile=@- https://0x0.st`
And send the link outputted by each command over here

**Comment by Alex-Zhangzhipeng** [comment]
Ok, i will try these , should i use DP only? or i use DMI or both?@ChefKissInc

**Comment by Alex-Zhangzhipeng** [comment]
sorry , maybe I misunderstand it , my pc is not a laptop @ChefKissInc 
![cd9aacc5c7492c576beb93d6525befc0](https://github.com/NootInc/NootedRed/assets/64397573/f00005be-7ae7-4bcf-8f40-06ddb7b66922)

**Comment by ChefKissInc** [comment]
@Alex-Zhangzhipeng So it’s just normal DisplayPort with the issue? Then the same thing but with your DisplayPort display working

**Comment by Alex-Zhangzhipeng** [comment]
sorry,I cannot understand it , my issue is that my DisplayPort cannot work fine when I change Windows to MacOS(windows is ok). HDMI is perfectly fine, DispalyPort is not. @ChefKissInc

**Comment by ChefKissInc** [comment]
@jalavoui what

**Comment by hhk02** [comment]
Hello i have the same laptop

**Comment by hhk02** [comment]
I installed sucessfully in Monterey with ASUS TUF GAMING FX505DT 

Works after the installation . First i installed the OS without the kext, and in the post installation i downloaded the kext and works.

**Comment by STEFnr1** [comment]
> I installed sucessfully in Monterey with ASUS TUF GAMING FX505DT
> 
> Works after the installation . First i installed the OS without the kext, and in the post installation i downloaded the kext and works.

Hello. Do you have the model with Ryzen 7? If yes, how did you make it work?

**Comment by hhk02** [comment]
> > I installed sucessfully in Monterey with ASUS TUF GAMING FX505DT
> > Works after the installation . First i installed the OS without the kext, and in the post installation i downloaded the kext and works.
> 
> Hello. Do you have the model with Ryzen 7? If yes, how did you make it work?

First I have installed macOS Monterey without the kext NootedRed once installed I have enabled the kext with the config.plist
Then restart the laptop and enable Secure Boot apply the changes in the BIOS when it reboots, disable Secure Boot again (This will enable the AMD GOP driver since if the Unknown Driver appears it could not continue loading the system) (I think that this thing that I have discovered disables the CSM) Make sure that the SMBIOS is MacBookPro16,3 otherwise the screen will be black.

If everything went well, when the system loads the screen will go black in a few seconds and then the apple will appear as a result the internal graphics will be running.

**Comment by truongvantuan** [comment]
> > > I installed sucessfully in Monterey with ASUS TUF GAMING FX505DT
> 
> > > Works after the installation . First i installed the OS without the kext, and in the post installation i downloaded the kext and works.
> 
> > 
> 
> > Hello. Do you have the model with Ryzen 7? If yes, how did you make it work?
> 
> Yes i have Ryzen 7 3570H with AMD Vega 10 too.
> 
> 
> 
> First I have installed macOS Monterey without the kext NootedRed once installed I have enabled the kext with the config.plist
> 
> Then restart the laptop and enable Secure Boot apply the changes in the BIOS when it reboots, disable Secure Boot again (This will enable the AMD GOP driver since if the Unknown Driver appears it could not continue loading the system) (I think that this thing that I have discovered disables the CSM) Make sure that the SMBIOS is MacBookPro16,3 otherwise the screen will be black.
> 
> 
> 
> If everything went well, when the system loads the screen will go black in a few seconds and then the apple will appear as a result the internal graphics will be running.
> 
> 
> 
> I recommend change Framebuffer to 512M to Auto (this increase to 2GB VRAM) for better experience.

VRAM set to Auto, then we can get 2GB vram? Is it possible?

**Comment by hhk02** [comment]
I tested and i have 2GB

**Comment by STEFnr1** [comment]
> > > > I installed sucessfully in Monterey with ASUS TUF GAMING FX505DT
> > 
> > 
> > > > Works after the installation . First i installed the OS without the kext, and in the post installation i downloaded the kext and works.
> > 
> > 
> > > 
> > 
> > 
> > > Hello. Do you have the model with Ryzen 7? If yes, how did you make it work?
> > 
> > 
> > Yes i have Ryzen 7 3570H with AMD Vega 10 too.
> > First I have installed macOS Monterey without the kext NootedRed once installed I have enabled the kext with the config.plist
> > Then restart the laptop and enable Secure Boot apply the changes in the BIOS when it reboots, disable Secure Boot again (This will enable the AMD GOP driver since if the Unknown Driver appears it could not continue loading the system) (I think that this thing that I have discovered disables the CSM) Make sure that the SMBIOS is MacBookPro16,3 otherwise the screen will be black.
> > If everything went well, when the system loads the screen will go black in a few seconds and then the apple will appear as a result the internal graphics will be running.
> > I recommend change Framebuffer to 512M to Auto (this increase to 2GB VRAM) for better experience.
> 
> VRAM set to Auto, then we can get 2GB vram? Is it possible?

Can confirm this is true. Setting the setting for the iGPU in BIOS to Auto makes the iGPU have 2GB of VRAM upon boot.

**Comment by truongvantuan** [comment]
> I tested and now i have 2GB

I have set Auto but just got 512MB VRAM, thanks bro

**Comment by hhk02** [comment]
> > I tested and now i have 2GB
> 
> I have set Auto but just got 512MB VRAM, thanks bro

![Captura de pantalla 2023-07-15 a las 23 30 30](https://github.com/NootInc/NootedRed/assets/80901734/2cc1a92e-9114-4c08-97c2-caccbfd1514e)

**Comment by VisualEhrmanntraut** [comment]
Additional information: The macOS kexts seemingly were based on Linux 5.12. However, we had one of the people run Linux 5.0, and that worked fine. We are even more lost than before :/

**Comment by VisualEhrmanntraut** [comment]
Caveat: Recovery still gets a black screen. Main OS works fine though.
